File: frontend/src/types/bracketOrder.ts
export enum OrderSide {
  BUY = "buy",
  SELL = "sell"
}

export enum EntryType {
  MARKET = "market",
  LIMIT = "limit"
}

export enum OrderStatus {
  PENDING = "pending",
  ACTIVE = "active",
  PARTIALLY_FILLED = "partially_filled",
  FILLED = "filled",
  CANCELLED = "cancelled",
  REJECTED = "rejected"
}

export interface TakeProfitLevel {
  price: number
  quantity: number
  order_id?: string
  filled_quantity?: number
}

export interface BracketOrderCreate {
  symbol: string
  side: OrderSide
  quantity: number
  
  // Entry configuration
  entry_type: EntryType
  entry_price?: number // Required for LIMIT, ignored for MARKET
  
  // Risk management (optional)
  stop_loss_price?: number
  
  // Profit targets (optional)
  take_profit_levels: TakeProfitLevel[]
}

export interface BracketOrderResponse {
  id: string
  symbol: string
  side: OrderSide
  quantity: number
  status: OrderStatus
  created_at: string
  
  // Entry configuration
  entry_type: EntryType
  entry_price?: number
  entry_order_id?: string
  entry_filled_quantity: number
  entry_average_price?: number
  
  // Stop loss
  stop_loss_price?: number
  stop_loss_order_id?: string
  
  // Take profit levels
  take_profit_levels: TakeProfitLevel[]
  
  // Calculated fields
  total_filled_quantity: number
  remaining_quantity: number
  total_pnl?: number
}

export interface BracketOrderFormData {
  symbol: string
  side: OrderSide
  quantity: string
  entry_type: EntryType
  entry_price: string
  stop_loss_price: string
  take_profit_1_price: string
  take_profit_1_quantity: string
  take_profit_2_price: string
  take_profit_2_quantity: string
}

File: backend/app/models/bracket_order.py
from pydantic import BaseModel
from typing import Optional, List
from decimal import Decimal
from enum import Enum
from datetime import datetime

class OrderSide(str, Enum):
    BUY = "buy"
    SELL = "sell"

class EntryType(str, Enum):
    MARKET = "market"
    LIMIT = "limit"

class OrderStatus(str, Enum):
    PENDING = "pending"
    ACTIVE = "active"
    PARTIALLY_FILLED = "partially_filled"
    FILLED = "filled"
    CANCELLED = "cancelled"
    REJECTED = "rejected"

class TakeProfitLevel(BaseModel):
    """Individual take profit level"""
    price: Decimal
    quantity: Decimal  # Absolute quantity (not percentage)
    order_id: Optional[str] = None
    filled_quantity: Decimal = Decimal("0")

class BracketOrderCreate(BaseModel):
    """Bracket order creation request"""
    symbol: str
    side: OrderSide
    quantity: Decimal
    
    # Entry configuration
    entry_type: EntryType
    entry_price: Optional[Decimal] = None  # Required for LIMIT, ignored for MARKET
    
    # Risk management (optional)
    stop_loss_price: Optional[Decimal] = None
    
    # Profit targets (optional)
    take_profit_levels: List[TakeProfitLevel] = []

class BracketOrderResponse(BaseModel):
    """Bracket order response"""
    id: str
    symbol: str
    side: OrderSide
    quantity: Decimal
    status: OrderStatus
    created_at: datetime
    
    # Entry configuration
    entry_type: EntryType
    entry_price: Optional[Decimal] = None
    entry_order_id: Optional[str] = None
    entry_filled_quantity: Decimal = Decimal("0")
    entry_average_price: Optional[Decimal] = None
    
    # Stop loss
    stop_loss_price: Optional[Decimal] = None
    stop_loss_order_id: Optional[str] = None
    
    # Take profit levels
    take_profit_levels: List[TakeProfitLevel] = []
    
    # Calculated fields
    total_filled_quantity: Decimal = Decimal("0")
    remaining_quantity: Decimal = Decimal("0")
    total_pnl: Optional[Decimal] = None
    
    class Config:
        json_encoders = {
            Decimal: str,
            datetime: lambda dt: dt.isoformat()
        }

class BracketOrderUpdate(BaseModel):
    """Update bracket order (only for pending orders)"""
    entry_price: Optional[Decimal] = None
    stop_loss_price: Optional[Decimal] = None
    take_profit_levels: Optional[List[TakeProfitLevel]] = None
    
    class Config:
        # Only include fields that are explicitly set
        fields = {
            'entry_price': {'exclude_unset': True},
            'stop_loss_price': {'exclude_unset': True},
            'take_profit_levels': {'exclude_unset': True}
        }

class BracketOrderValidationError(Exception):
    """Custom exception for bracket order validation errors"""
    pass